src/Data/Nat/Binary/Minus.agda

```diff
@@ -0,0 +1,239 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Subtraction on Bin and some of its property proofs.
+------------------------------------------------------------------------
+
+{-# OPTIONS --without-K --safe #-}
+
+module Data.Nat.Binary.Minus where
+
+open import Data.Nat.Binary.Base
+open import Data.Nat.Binary.Properties
+open import Data.Empty using (⊥)
+open import Data.Nat as ℕ using (ℕ)
+import Data.Nat.Properties as ℕₚ
+open import Data.Product using (_×_; _,_; proj₁; proj₂; ∃)
+open import Data.Sum using (inj₁; inj₂)
+open import Function using (_∘_; _$_)
+open import Relation.Binary using (Tri; _Preserves_⟶_; _Preserves₂_⟶_⟶_)
+open import Relation.Binary.PropositionalEquality as P using
+  (_≡_; _≢_; _≗_; refl; sym; trans; cong; cong₂; resp₂; subst)
+import Relation.Binary.Reasoning.Base.Triple as InequalityReasoning
+open import Relation.Nullary using (Dec; yes; no)
+open import Relation.Nullary.Negation using (contradiction)
+open import Algebra.FunctionProperties {A = ℕᵇ} _≡_
+open import Algebra.Properties.CommutativeSemigroup +-commSemigroup
+     using (xy∙z≈y∙xz; x∙yz≈y∙xz; xy∙zu≈xz∙yu)
+
+private
+  module IneqReasoning = InequalityReasoning {A = ℕᵇ} {_≈_ = _≡_} ≤-isPreorder
+                                   <-trans (resp₂ _<_) <⇒≤ <-≤-trans ≤-<-trans
+open ℕᵇ
+open Tri
+open _<_
+
+Minus : ℕᵇ → ℕᵇ → Set                                 -- result of subtraction
+Minus x y =  ℕᵇ × Tri (x < y) (x ≡ y) (x > y)
+
+minus :  ∀ x y → Minus x y                            -- extended subtraction
+minus zero     zero       =  (0ᵇ ,       tri-≈ refl)
+minus zero     (2[1+ _ ]) =  (0ᵇ ,       tri-< 0<even)
+minus zero     1+[2 _ ]   =  (0ᵇ ,       tri-< 0<odd)
+minus 2[1+ x ] zero       =  (2[1+ x ] , tri-> 0<even)
+minus 2[1+ x ] 2[1+ y ]   with minus x y
+... | (d , cmp) = (double d , even<even-mono-cmp cmp)
+minus 2[1+ x ] 1+[2 y ]  with minus x y
+... | (_ , tri< x<y _   _)   =  (0ᵇ  , tri-< (even<odd x<y))
+... | (d , tri≈ _   x≡y _)   =  (1+[2 d ] , tri-> (odd<even (inj₂ (sym x≡y))))
+... | (d , tri> _   _   x>y) =  (1+[2 d ] , tri-> (odd<even (inj₁ x>y)))
+minus 1+[2 x ] zero      =  (1+[2 x ] , tri-> 0<odd)
+minus 1+[2 x ] 2[1+ y ]  with minus x y
+... | (_ , tri< x<y _  _)   =  (0ᵇ , tri-< (odd<even (inj₁ x<y)))
+... | (_ , tri≈ _   eq _)   =  (0ᵇ , tri-< (odd<even (inj₂ eq)))
+... | (d , tri> _   _  x>y) =  (pred (double d) , tri-> (even<odd x>y))
+minus 1+[2 x ] 1+[2 y ]  with minus x y
+... | (d , cmp) =  (double d , odd<odd-mono-cmp cmp)
+
+infixl 6 _∸_
+
+_∸_ : Op₂ ℕᵇ              -- traditional subtraction
+_∸_ x =  proj₁ ∘ minus x
```

Comment:
I think your definition of minus is unnecessarily complicated. I would change it for the much shorter definition:
```
_∸_ :  ℕᵇ → ℕᵇ → ℕᵇ
zero     ∸ _        = 0ᵇ
x        ∸ zero     = x
2[1+ x ] ∸ 2[1+ y ] = double (x ∸ y)
1+[2 x ] ∸ 1+[2 y ] = double (x ∸ y)
2[1+ x ] ∸ 1+[2 y ] with x <? y
... | yes _  = 0ᵇ
... | no  _  = 1+[2 (x ∸ y) ]
1+[2 x ] ∸ 2[1+ y ] with x ≤? y
... | yes _  = 0ᵇ
... | no  _  = pred (double (x ∸ y))
```
This cuts down the number of cases significantly, making subsequent proofs shorter. The proof components from the removed trichotomous component can easily be recovered from the proofs from `yes` and `no`

Reply:
Your statement  for ``2[1+ x ] ∸ 1+[2 y ]`` computes ``x <? y`` and then computes ``x ∸ y`` by new.
And in my program, computing ``minus 2[1+ x ] 1+[2 y ]`` applies ``minus x y``, which also gives the result of  ``<-cmp x  y`` and  ``x <? y``.  This latter result is formed for free as an addition to the difference denoted ``d`` in the code.
My format for ``minus`` is chosen so because I think that a natural way of computing ``x ∸ y``   
includes computing ``x <? y``. So that there is no need to apply ``x <? y`` separately.
I expect that my ``minus`` is about 3/2 times faster than your ``∸``.
But I am not sure.

Reply:
You may very well be right, but I'm afraid I don't quite see it myself why your version is more efficient than mine. Yes you're right that mine does effectively calculate the difference multiple times. However doesn't yours also do that with all the translations of `tri-<`, `tri->` etc?

I would be interested in seeing the timings...

Reply:
I doubt. 
I shall look into this after may be a week. I need a time out.

Reply:
I tried a certain test, and your version for ``∸`` has occurred faster, I do not understand, why.

All right, let us choose your version.
Let us delay this performance analysis, it is better to advance with the project. 
I have changed the whole Minus.agda, accommodating  it to your version of ``∸`.
But I do not recall the situation with this PR. Probably  _master_ has changed greatly.
So, I am going to do the following.
* To save your last notes on this PR,
* To delete this my branch from GitHub.
* To start a new branch from _master_ with a PR with Minus.agda, together with several 
   small evident items of which you suggested me to do a PR.
   Because everything except Minus.agda is already in _master_. Right?

Reply:
> Because everything except Minus.agda is already in master. Right?

That is correct.

But you don't necessarily have to delete the branch and start again. You could [pull the changes down from the library and merge them in](https://stackoverflow.com/questions/34656523/git-update-local-branch-with-remote-master/34656923)?

Reply:
I want to replace my local branch with the copy of remote _master_. 
Only this, to start with.
Is this by
```
$ cd .../agda-stdlib-fork        (my local, where CHANGELOG.md is)
$ git checkout master
$ git pull
```
? 
On the Web, it is also added 
```
$ git checkout <your-branch>
$ git merge master
```
Is this needed? If yes, then what need I set for <your-branch> ?  Is it '.' ?

Reply:
> I want to replace my local branch with the copy of remote master.

That's not quite what you want to do. You want merge the remote master branch with your master branch.

After your first two commands I would run `git merge origin master`

Reply:
I followed this. It reports "Already updated" (translated from Russian).
Now I am going to commit a PR with Nat.Binary.Minus